### PR TITLE
[wip] authn+authz

### PR DESCRIPTION
**Pull request by clint-tseng**

### CODE REVIEW

**Comment by clint-tseng** [comment]
Okay, please to review @lognaturel @matthew-white @ggalmazor. Commit messages are highly descriptive.

Explicitly open questions:
1. There aren't yet tests; I have ideas on approaches here (test the raw db queries more explicitly; test the higher-order queries logically; do integration tests at the HTTP level (but what to do about the database?)), but I'd like to hear suggestions now that the code is more concrete.
2. Some of the linting things are maybe a bit insane (`no-use-before-define`).
3. `.transacting` vs `.transacting()` are currently not consistent; would like to hear which is preferred and I will push them all in that direction.

Downriver questions that don't prevent this PR from merging:
1. How exactly we want to expose and machine the rudimentary perms settings we've discussed for MVP (eg "anybody can register an account" vs "admins must create an account"; "any account holder can create a new form" vs "admins must create a new form"; etc).
    a. Option A: the backend has no idea about them and the frontend simply looks for and creates/deletes the appropriate grants records explicitly, or
    b. Option B: the backend understands and explicitly exposes a simplified `/settings` endpoint that provides some of these concepts as higher-level booleans.
    c. The answer here should keep in mind the down-the-road state of things, where these rudimentary options work alongside a more flexible permission assignment UI.
2. How to do session refresh; right now all sessions last for exactly 24 hours with no refresh.
    a. Option A: implicitly kick expiry out to 24 hours from each use. (pros: simple; cons: `GET`s trigger database writes).
    b. Option B: use OAuth2's refresh token mechanism. (pros: explicit, standardized, we'll need it anyway; cons: more homework up front (n.b. I don't mind doing this homework)).
    c. Option C: ignore the problem for now.
3. We'll need HTTPS to protect plaintext passwords in transport.
4. The mechanism whereby the frontend is informed of the current sessionholder's rights on various things, so that it can appropriately provide the correct verbs/grey out inaccessible operations, etc.
    a. Option A: `GET`ted objects contain a `grants` key containing an array of available verbs.
    b. Option B: A separate API endpoint may be queried to retrieve said array of verbs. (We will likely need this endpoint anyway, but the question stands.)

More options exist for each of these; these are just the obvious paths in my mind.

**Comment by clint-tseng** [comment]
Okay, now basic Form POST/GET are in there too since they were incredibly easy to add.

File: lib/resources/forms.js [hunk 1]
```diff
@@ -1,20 +1,17 @@
...
-    )
-  ));
-  */
+module.exports = (service, { Actee, Form }) => {
+  service.post('/forms', endpoint(({ body, session }) =>
+    session.transacting
+      .canOrReject('create', Actee.species('form'))
+      .then(() => Form.fromXml(body))
+      .then((form) => form.create())
+      .then((savedForm) => (session.actor.isDefined()
```
Comment (clint-tseng) [inline lib/resources/forms.js]: Lacking pattern-matching, one wonders whether this can be made smoother.

File: lib/model/instance/session.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    return new Session({
+      actorId: actor.id,
+      token: randomBytes(48).toString('base64'),
+      expires
+    });
+  }
+
+  static getByBearerToken(token) {
+    return sessions.getByBearerToken(token)
+      // TODO: awkward commuting homework.
```
Comment (clint-tseng) [inline lib/model/instance/session.js]: Need to go from Option[T] to Promise[Option[T]], which is awkward with the current machinery.

File: lib/model/instance/instance.js [hunk 3]
```diff
@@ -1,12 +1,22 @@
...
 // methods sprinkled in to ease the readability of common tasks.
 //
 // InstanceBase is the only common superclass to them all.
 class InstanceBase {
   forCreate() { return this; }
-  forSerialize() { return this; }
+  forApi() { return this; }
 
   // Returns a new copy of this instance with merged properties.
   with(data) {
```
[the remark this answers is not in the document]
Comment (clint-tseng) [reply lib/model/instance/instance.js]: ES5 removed a lot of limitations with reserved keywords; they are only illegal in cases where the lexer could genuinely be confused. If you have another verb to suggest that is as descriptive as `with` I am happy to substitute, though.

File: lib/resources/users.js [hunk 4]
```diff
@@ -1,14 +1,39 @@
...
-
-  service.post('/users', endpoint(({ body }) =>
-    User.fromSerialize(body).create()));
+module.exports = (service, { Actee, User }) => {
 
   // Gets full details of a user by actor id.
-  service.get('/users/:id', endpoint(({ params }) =>
-    User.getByActorId(params.id).then((user) => user.withActor())));
+  // TODO: probably hide email addresses unless admin/self?
+  service.get('/users/:id', endpoint(({ params }) => User.getByActorId(params.id)));
```
[the remark this answers is not in the document]
Comment (clint-tseng) [reply lib/resources/users.js]: Okay; sounds like future work though.

[the remark this answers is not in the document]
Comment (clint-tseng) [reply lib/resources/users.js]: #33

**Comment by clint-tseng** [comment]
I hear what you're saying on "easy"/"just"/etc, but without such words it is difficult to express the justification for punting certain things down the road.

**Comment by clint-tseng** [comment]
I'm going to merge this just to simplify the cross-repository stuff and start working on subbranches but please don't let that stop anybody from continuing to review and comment.

**Comment by ggalmazor** [comment]
Hi everyone!

I've reviewed the commit messages and I've also downloaded your branch and checked the code and I can't see any important stuff that we should have to take care of right now. I feel that with this PR, Jubilant is taking a good momentum. Huge step forward here, congratulations! :)

**Comment by ggalmazor** [comment]
Regarding the tests, I can't really add much to what I said in #29 at this moment. Maybe after you develop your ideas I can be more helpful. As I mentioned in Slack, I'll be focusing on Aggregate and Briefcase but I can still help with specific problems.

**Pull request merged**